adafruit_requests.py

```diff
@@ -49,263 +50,537 @@
+            self._last_response.close()
+            self._last_response = None
+
+        socket = self._get_socket(host, port, proto, timeout=timeout)
+        socket.send(
+            b"%s /%s HTTP/1.1\r\n" % (bytes(method, "utf-8"), bytes(path, "utf-8"))
         )
         if "Host" not in headers:
-            sock.send(b"Host: %s\r\n" % bytes(host, "utf-8"))
+            socket.send(b"Host: %s\r\n" % bytes(host, "utf-8"))
         if "User-Agent" not in headers:
-            sock.send(b"User-Agent: Adafruit CircuitPython\r\n")
+            socket.send(b"User-Agent: Adafruit CircuitPython\r\n")
         # Iterate over keys to avoid tuple alloc
         for k in headers:
-            sock.send(k.encode())
-            sock.send(b": ")
-            sock.send(headers[k].encode())
-            sock.send(b"\r\n")
+            socket.send(k.encode())
+            socket.send(b": ")
+            socket.send(headers[k].encode())
+            socket.send(b"\r\n")
         if json is not None:
             assert data is None
             # pylint: disable=import-outside-toplevel
             try:
                 import json as json_module
             except ImportError:
                 import ujson as json_module
-            # pylint: enable=import-outside-toplevel
             data = json_module.dumps(json)
-            sock.send(b"Content-Type: application/json\r\n")
+            socket.send(b"Content-Type: application/json\r\n")
         if data:
             if isinstance(data, dict):
-                sock.send(b"Content-Type: application/x-www-form-urlencoded\r\n")
+                socket.send(b"Content-Type: application/x-www-form-urlencoded\r\n")
                 _post_data = ""
                 for k in data:
                     _post_data = "{}&{}={}".format(_post_data, k, data[k])
                 data = _post_data[1:]
-            sock.send(b"Content-Length: %d\r\n" % len(data))
-        sock.send(b"\r\n")
+            socket.send(b"Content-Length: %d\r\n" % len(data))
+        socket.send(b"\r\n")
         if data:
             if isinstance(data, bytearray):
-                sock.send(bytes(data))
+                socket.send(bytes(data))
             else:
-                sock.send(bytes(data, "utf-8"))
-
-        line = sock.readline()
-        # print(line)
-        line = line.split(None, 2)
-        status = int(line[1])
-        reason = ""
-        if len(line) > 2:
-            reason = line[2].rstrip()
-        resp.headers = parse_headers(sock)
-        if resp.headers.get("transfer-encoding"):
-            if "chunked" in resp.headers.get("transfer-encoding"):
-                raise ValueError("Unsupported " + resp.headers.get("transfer-encoding"))
-        elif resp.headers.get("location") and not 200 <= status <= 299:
+                socket.send(bytes(data, "utf-8"))
+
+        resp = Response(socket, self)  # our response
+        if "location" in resp.headers and not 200 <= resp.status_code <= 299:
```

Comment:
May want to be more explicit on this `NotImplementedError`?  Status codes outside this range aren't just 3xx redirection. Could be be 4xx client errors or 5xx server errors.

Reply:
I made the if just apply to redirects.